README
LogSender
=========
Version: 0.1.1
Date: 2018-10-15

Send files from a set of local directories to one or several remote machines
via ftp and take care to only send them once.


Enjoy!

Tobias Oetiker <tobi@oetiker.ch>


CHANGES
0.4.0 2023-05-12 17:43:02 +0200 Tobias Oetiker <tobi@oetiker.ch>

 - allow sftp urls of the form sftp://host[:port]
   note, this relies on ssh keys already being setup for the user that is running
   logsender. Put the ssh configuration into ~/.ssh/config

0.2.5 2016-09-23  Tobias Oetiker (tobi@oetiker.ch)
- properly release filehandles

0.0.0 2016-03-22  Tobias Oetiker (tobi@oetiker.ch)

- started project


dist.sh
#!/bin/bash
V=$(cat VERSION)
P=logsender
set -eo pipefail
if ssh oepdown@freddielx -- test -f public_html/hin/${P}-${V}.tar.gz; then
   echo version ${V} this version has already been published
   exit 1
fi

echo ${V} `date +"%Y-%m-%d %H:%M:%S %z"` `git config user.name` '<'`git config user.email`'>' >> CHANGES.new
echo >> CHANGES.new
echo ' -' >> CHANGES.new
echo >> CHANGES.new
cat CHANGES >> CHANGES.new && mv CHANGES.new CHANGES
$EDITOR CHANGES
rm -f config.status
./bootstrap
./configure
make
make test | tee ${P}-${V}.test-output.txt
make dist
docker build --build-arg V=${V} --pull --tag ${P}:${V} .
docker save ${P}:${V} --output ${P}-${V}.docker
pixz ${P}-${V}.docker
chmod 644 ${P}-${V}.docker.xz
cat ${P}-${V}.tar.gz distsecret | sha512sum > ${P}-${V}.tar.gz.sum
scp ${P}-${V}.docker.xz ${P}-${V}.tar.gz* ${P}-${V}.test-output.txt oepdown@freddielx:public_html/hin/


lib/LogSender.pm
package LogSender;

use Mojo::Base 'Mojolicious';
use LogSender::Config;

our $VERSION = "0.0.0";

=head1 NAME

LogSender - Application Class

=head1 SYNOPSIS

 logsender.pl COMMAND

=head1 DESCRIPTION

Configure the mojolicious engine to run our application logic

=cut

=head1 ATTRIBUTES

LogFetcher has all the attributes of L<Mojolicious> plus:

=cut

=head2 config

use our own plugin directory and our own configuration file:

=cut

has config => sub {
    my $app = shift;
    LogSender::Config->new(
        app => $app,
        file => $ENV{LOGSENDER_CFG} || $app->home->rel_file('etc/logsender.cfg' )
    );
};


sub startup {
    my $app = shift;
    @{$app->commands->namespaces} = (__PACKAGE__.'::Command');
    $app->commands->message("Usage:\n\n".$app->commands->extract_usage."\nCommands:\n\n");
}

1;

=head1 COPYRIGHT

Copyright (c) 2016 by OETIKER+PARTNER AG. All rights reserved.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=cut


lib/LogSender/Command/send.pm
package LogSender::Command::send;
use Mojo::Base 'Mojolicious::Command';
use Getopt::Long 2.25 qw(:config posix_default no_ignore_case);

#use Test::Mock::Net::FTP;
#Test::Mock::Net::FTP::mock_prepare(engelberg => {oetiker => {password => 'gugus',dir => [ "/home/oetiker/scratch/log/remote_ftp/", "/" ]},},);
#my $FTP = 'Test::Mock::Net::FTP';

use Net::FTP;
use Net::SFTP::Foreign;

use Mojo::URL;
use Mojo::Util qw(dumper);
use File::Basename;
use Time::HiRes qw(gettimeofday);
use POSIX qw(strftime);
use File::Spec;
use IO::Uncompress::Gunzip qw(gunzip $GunzipError);
use File::Temp;

=head1 NAME

LogSender::Command::send - LogSender send tool

=head1 SYNOPSIS

 ./logsender.pl send

=head1 DESCRIPTION

Run logsender to push local logfiles to the servers. The url parameter can contain ftp or sftp urls.
Note that the sftp functionality uses the localy installed ssh client to do the transfer. This means
that the user running logsender must be able to connect to the remote host without password. This
can be achieved by using ssh keys. These keys must be generated for the user running logsender and
the public key must be added to the remote users authorized_keys file. This all happens in the $HOME/.ssh
directory of the user running logsender.

=cut

has description => <<'EOF';
copy logfiles to a remote machine via ftp and keep track of what has been transfered
EOF

has usage => <<"EOF";
usage: $0 send [OPTIONS]

These options are available:

  --verbose   talk about your work
  --debug     talk even more
  --noaction  ONLY talk do not act

Operations of logsender send are configured in the logsend.cfg file. Use
the LOGSENDER_CFG environment variable to point to an alternate location
for this file.

EOF

my %opt;

has log => sub { shift->app->log };

has cfg => sub { shift->app->config->cfgHash };


sub run {
    my $self   = shift;
    local @ARGV = @_ if scalar @_;
    GetOptions(\%opt,'noaction|no-action|n', 'verbose|v','debug|d');
    if ($opt{verbose} or $opt{noaction}){
        $self->log->level('debug');
        $self->log->handle(\*STDOUT);
    }

    while(1){
        my $now = time;
        $self->action;
        my $sleep  = $self->cfg->{GENERAL}{logCheckInterval}-(time-$now);
        $self->log->debug("sleeping for $sleep s ...");
        sleep $sleep;
    }
}

sub action {
    my $self = shift;
    my $app = $self->app;
    my $cfg = $self->cfg;

    my $now = time;
    HOST:
    for my $host (@{$cfg->{HOSTS}}){
        my $suffix = $host->{transferSuffix};
        my $url = Mojo::URL->new($host->{url});
        $self->log->debug("Connect to $host->{url}");

        my $ftp;
        for ($url->scheme) {
            /^ftp$/ && do {
                $ftp = Net::FTP->new($url->host,Port=>$url->port||21,Passive=>1) or do {
                    $self->log->warn("Skipping ".$host->{url}.": ".$@);
                    next HOST;
                };
                $ftp->login($url->userinfo ? (split /:/,$url->userinfo) : () ) or do {
                    $self->log->error("login failed: " . $ftp->code() . ": " . $ftp->message());
                    next HOST;
                };

                $ftp->binary() or do {
                    $self->log->error("config binary failed: " . $ftp->code() . ": " . $ftp->message());
                    next HOST;
                };
                last;
            };
            /^sftp$/ && do {
                $ftp = Net::SFTP::Foreign->new(
                    host=>$url->host,
                    $url->port ? (port => $url->port) : (),
                    more => [ qw(-o PreferredAuthentications=publickey), ($opt{debug} ? ('-v') : ())]
                ) or do {

                    $self->log->error("ssh connection failed: " . $ftp->error() . ": " . $ftp->status());
                    next HOST;
                };
                last;
            };
        }
        $self->log->debug("Connected to $host->{url}");
        for my $file (@{$host->{FILES}}){
            my %glob;
            for my $gen ( 0..$cfg->{GENERAL}{logCheckPastIntervals}){
                $glob{strftime($file->{globPattern},localtime($now-($gen*$cfg->{GENERAL}{logCheckInterval})))} = 1;
            }
            # just try makeing the destination directory
            # if it failed, put will fail, so we don't check here.
            if ($url->scheme eq 'ftp') {
                $ftp->mkdir($file->{destinationDir},1) unless $opt{noaction};
            }
            elsif ($url->scheme eq 'sftp') {
                $ftp->mkpath($file->{destinationDir}) unless $opt{noaction};
            }
            for my $pattern (keys %glob){
                for my $path (glob $pattern){
                    next if $path =~ m/\Q${suffix}\E$/;
                    my $basename = basename($path);
                    next if $file->{skipFile} and strftime($file->{skipFile},localtime(time)) eq $basename;
                    next if $file->{stopFile} and $basename le $file->{stopFile};
                    next if not -f $path;
                    # only touch files which have not changed for 10 minutes
                    next if time - [stat($path)]->[10] < 600;
                    my $mark = $path.$suffix;
                    next if -e $mark;
                    my $start = gettimeofday();
                    if ($opt{noaction}){
                        $self->log->debug("skipping put $path (noaction mode)");
                        next;
                    }
                    my $src = $path;
                    my $fh;
                    if ($host->{'gunzip'} eq 'yes' and $basename =~ s/\.gz$//){
                        $fh = File::Temp->new( TEMPLATE => 'logsenderXXXXX', TMPDIR => 1 );
                        gunzip $path,$fh or do {
                            $self->log->error("gunzip $path failed: $GunzipError");
                        };
                        $fh->flush;
                        $fh->sync;
                        $src=$fh->filename;
                        $self->log->debug("gunzip $path to $src");
                    }
                    open my $touch,'>',$mark or do {
                        $self->log->error("aborting transfer of $path - creating $mark failed: $!");
                        next;
                    };
                    $self->log->debug("sending $src to $file->{destinationDir}/$basename");
                    $ftp->put($src,File::Spec->catfile($file->{destinationDir},$basename)) or do {
                        if ($url->scheme eq 'ftp') {
                            $self->log->error("aborting transfer of $path: " . $ftp->code() . ": " . $ftp->message());
                        } else {
                            $self->log->error("aborting transfer of $path: " . $ftp->error() . ": " . $ftp->status());
                        }
                        unlink $mark;
                        next;
                    };
                    my $size = -s $src;
                    my $end = gettimeofday();
                    $self->log->debug("$src transferred $size Bytes @ ".sprintf("%.1f MByte/s",($size/(1024*1024))/($end-$start)));
                }
            }
        }
    }
}

1;
__END__

=back

=head1 COPYRIGHT

Copyright (c) 2016 by OETIKER+PARTNER AG. All rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2011-05-30 to 1.0 first version
 2023-05-12 to 1.2 added sftp support

=cut

# Emacs Configuration
#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et


lib/LogSender/Config.pm
package LogSender::Config;

use Mojo::Base -base;
use Mojo::JSON qw(decode_json);
use Mojo::File;
use Mojo::Exception;
use Data::Processor;
use Data::Processor::ValidatorFactory;

=head1 NAME

LogSender::Config - the Config access  class

=head1 SYNOPSIS

 use LogSender::Config;
 my $conf = LogFetcher::Config->new(app=>$app,file=>'config.json');
 my $hash = $conf->cfgHash;
 print $hash->{GENERAL}{value};

=head1 DESCRIPTION

Load and preprocess a configuration file in json format.

=head1 ATTRIBUTES

All the attributes from L<Mojo::Base> as well as:

=head2 app

pointing to the app

=cut

has 'app';

has validatorFactory => sub {
    Data::Processor::ValidatorFactory->new;
};

has validator => sub {
    Data::Processor->new(shift->schema);
};

=head2 file

the path of the config file

=cut

has 'file';


=head2 SCHEMA

the flattened content of the config file

=cut

my $CONSTANT_RE = '[_A-Z]+';

has schema => sub {
    my $self = shift;
    my $vf = $self->validatorFactory;
    my $string = $vf->rx('^.*$','expected a string');
    my $url = $vf->rx('^(ftp://[^:]+:[^@]+|sftp://[^\s@:]+(:\d+)?)$','expected [s]ftp url: ftp://user[:pass]@host or sftp://host[:port]');
    my $integer = $vf->rx('^\d+$','expected an integer');

    return {
        GENERAL => {
            description => 'general settings',
            members => {
                logFile => {
                    validator => $vf->file('>>','writing'),
                    description => 'absolute path to log file',
                },
                logLevel => {
                    validator => $vf->rx('(?:debug|info|warn|error|fatal)','Pick a logLevel of debug, info, warn, error or fatal'),
                    description => 'mojo log level - debug, info, warn, error or fatal'
                },
                logCheckInterval => {
                    description => 'log check interval in seconds',
                    validator => $integer,
                },
                logCheckPastIntervals => {
                    description => 'how many intervals into the past to check',
                    validator => $integer,
                },
                timeout => {
                    description => 'how long to wait for transfer to timeout',
                    validator => $integer,
                    default => 5,
                },
            },
        },
        CONSTANTS => {
            description => 'define constants fo be used in globPattern properties.',
            optional => 1,
            members => {
                $CONSTANT_RE => {
                    regex => 1,
                    description => 'value of the constant',
                    validator => $string,
                }
            }

        },
        HOSTS => {
            description => 'where does our data go to.',
            array => 1,
            members => {
                url => {
                    description => 'ftp url with username and password',
                    example => 'ftp://myuser:mypass@myhost',
                    validator => $url
                },
                transferSuffix => {
                    description => 'When a file has been successfully transfered, the LogSender will touch the filename with the transferSuffix appended thus preventing any further transfers of the file.',
                    validator => $string
                },
                gunzip => {
                    description => 'gunzip any file ending in .gz prior to transfer?',
                    validator => $vf->rx('yes|no','expected yes or no')
                },
                FILES => {
                    description => 'a map of globs on the remote machine',
                    array => 1,
                    members => {
                        destinationDir => {
                            description => 'destination directory on the ftp server',
                            validator => $string,
                            optional => 1
                        },
                        globPattern => {
                            description => 'a glob pattern to find all logfiles to send you can use ${CONSTANTS} and strftime placeholders',
                            example => '/var/log/archive/%Y/%m/%d/messages-*.gz',
                            validator => $string
                        },
                        skipFile => {
                            description => 'a string to match against the filename. If it matches the file will be skipped. You can use ${CONSTANTS} and strftime placeholders. It operates on localtime.',
                            example => 'access_log.%Y-%m-%d',
                            validator => $string,
                            optional => 1
                        },
                        stopFile => {
                            optional => 1,
                            description => 'If defined, no filename C<le> this one will be considered for transfer.',
                            validator => $string
                        },
                    }
                }
            }
        }
    };
};

=head2 rawCfg

raw config

=cut

has rawCfg => sub {
    my $self = shift;
    $self->n3kCommon->loadJSONCfg($self->file);
};


=head2 cfgHash

access the config hash

=cut

has cfgHash => sub {
    my $self = shift;
    my $cfg = $self->loadJSONCfg($self->file);
    my $validator = $self->validator;
    my $hasErrors;
    my $err = $validator->validate($cfg);
    for ($err->as_array){
        warn "$_\n";
        $hasErrors = 1;
    }
    die "Can't continue with config errors\n" if $hasErrors;
    # we need to set this real early to catch all the info in the logfile.
    $self->app->log->path($cfg->{GENERAL}{logFile});
    $self->app->log->level($cfg->{GENERAL}{logLevel});
    if (my $const = $cfg->{CONSTANTS}){
        my $CONST_MATCH = join('|',keys %$const);
        for my $host (@{$cfg->{HOSTS}}){
            for my $logFile (@{$host->{FILES}}){
                for my $key (qw(globPattern skipFile)){
                    next unless defined $logFile->{$key};
                    $logFile->{$key} =~ s/\$\{($CONST_MATCH)\}/$const->{$1}/g;
                }
            }
        }
        delete $cfg->{CONSTANTS};
    }
    return $cfg;
};

=head1 METHODS

All the methods of L<Mojo::Base> as well as:

=head2 loadJSONCfg(file)

Load the given config, sending error messages to stdout and igonring /// lines as comments

=cut

sub loadJSONCfg {
    my $self = shift;
    my $file = shift;
    my $json = Mojo::File->new($file)->slurp;
    $json =~ s{^\s*//.*}{}gm;

    my $raw_cfg = eval { decode_json($json) };
    if ($@){
        if ($@ =~ /(.+?) at line (\d+), offset (\d+)/){
            my $warning = $1;
            my $line = $2;
            my $offset = $3;
            open my $json, '<', $file;
            my $c =0;
            warn "Reading ".$file."\n";
            warn "$warning\n\n";
            while (<$json>){
                chomp;
                $c++;
                if ($c == $line){
                    warn ">-".('-' x $offset).'.'."  line $line\n";
                    warn "  $_\n";
                    warn ">-".('-' x $offset).'^'."\n";
                }
                elsif ($c+3 > $line and $c-3 < $line){
                    warn "  $_\n";
                }
            }
            warn "\n";
            exit 1;
        }
        else {
            Mojo::Exception->throw("Reading ".$file.': '.$@);
        }
    }
    return $raw_cfg;
}


1;

__END__

=head1 COPYRIGHT

Copyright (c) 2016 by OETIKER+PARTNER AG. All rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2016-03-22 to 0.0 first version

=cut

# Emacs Configuration
#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et
